util-cache/src/main/scala/sbt/util/ActionCache.scala

```diff
@@ -69,38 +64,82 @@ object ActionCache:
         val newOutputs = Vector(valueFile) ++ outputs.toVector
         store.put(UpdateActionResultRequest(input, newOutputs, exitCode = 0)) match
           case Right(cachedResult) =>
-            syncBlobs(cachedResult.outputFiles)
+            store.syncBlobs(cachedResult.outputFiles, outputDirectory)
             result
           case Left(e) => throw e
 
+    get(key, codeContentHash, extraHash, tags, config) match
+      case Some(value) => value
+      case None        => organicTask
+  end cache
+
+  /**
+   * Retrieves the cached value.
+   */
+  def get[I: HashWriter, O: JsonFormat: ClassTag](
+      key: I,
+      codeContentHash: Digest,
+      extraHash: Digest,
+      tags: List[CacheLevelTag],
+      config: BuildWideCacheConfiguration,
+  ): Option[O] =
+    import config.store
     def valueFromStr(str: String, origin: Option[String]): O =
-      cacheEventLog.append(ActionCacheEvent.Found(origin.getOrElse("unknown")))
+      config.cacheEventLog.append(ActionCacheEvent.Found(origin.getOrElse("unknown")))
       val json = Parser.parseUnsafe(str)
       Converter.fromJsonUnsafe[O](json)
-
-    def syncBlobs(refs: Seq[HashedVirtualFileRef]): Seq[Path] =
-      store.syncBlobs(refs, config.outputDirectory)
-
-    val getRequest =
-      GetActionResultRequest(input, inlineStdout = false, inlineStderr = false, Vector(valuePath))
-    store.get(getRequest) match
+    findActionResult(key, codeContentHash, extraHash, config) {
       case Right(result) =>
         // some protocol can embed values into the result
         result.contents.headOption match
           case Some(head) =>
-            syncBlobs(result.outputFiles)
+            store.syncBlobs(result.outputFiles, config.outputDirectory)
             val str = String(head.array(), StandardCharsets.UTF_8)
-            valueFromStr(str, result.origin)
+            Some(valueFromStr(str, result.origin))
           case _ =>
-            val paths = syncBlobs(result.outputFiles)
-            if paths.isEmpty then organicTask
-            else valueFromStr(IO.read(paths.head.toFile()), result.origin)
-      case Left(_) => organicTask
-  end cache
+            val paths = store.syncBlobs(result.outputFiles, config.outputDirectory)
+            if paths.isEmpty then None
+            else Some(valueFromStr(IO.read(paths.head.toFile()), result.origin))
+      case Left(_) => None
+    }
+
+  /**
+   * Checks if the ActionResult exists in the cache.
+   */
+  def exists[I: HashWriter](
+      key: I,
+      codeContentHash: Digest,
+      extraHash: Digest,
+      config: BuildWideCacheConfiguration,
+  ): Boolean =
+    findActionResult(key, codeContentHash, extraHash, config) {
+      case Right(_) => true
+      case Left(_)  => false
+    }
+
+  inline private[sbt] def findActionResult[I: HashWriter, O](
+      key: I,
+      codeContentHash: Digest,
+      extraHash: Digest,
+      config: BuildWideCacheConfiguration,
+  )(f: Either[Throwable, ActionResult] => O) =
```

Comment:
Why passing a lambda here? It seems the caller just need to match on the `Either` of the action result.

```suggestion
  inline private[sbt] def findActionResult[I: HashWriter, O](
      key: I,
      codeContentHash: Digest,
      extraHash: Digest,
      config: BuildWideCacheConfiguration,
  ): Either[Throwable, ActionResult] =
```

Reply:
Yea, you're right. For some reason, I thought letting it accept a function would make the other functions make nicer to write, but I guess it won't make much difference.

Reply:
Addressed in 291778086d6a921b099c337d07bffc69201bd655